### ISSUE TITLE
Memory leak related to accumulation of ParseExceptions in `all_exceptions`

### ISSUE THREAD

**Issue by erp12** [mozilla/moz-sql-parser#79]
I have been tracking down memory issues in my application and stumbled upon a possible issue with the `DEBUG` option for the parser. 

Brief background on my application, it is an evolutionary algorithm that is essentially one big loop. There should be almost no accumulation of data between each iteration of the loop and GC should be able to clean up everything that the previous iteration was manipulating. Below are some counts for different objections I see each iteration of the loop. Clearly some types of objects are sticking around between iterations.

![object_growth](https://user-images.githubusercontent.com/5595494/54885083-df2a0780-4e4e-11e9-9d4c-79c8d7f037fe.png)

(Horizontal axis is loop iteration)

Definitely a lot of pyparsing ParseExceptions, dicts, and tuples being kept around. I used the `objgraph` package to check what modules were referencing these objects. I checked 10 random instances per loop iteration, and they all pointed to the `moz_sql_parser` module via lists within the `all_exceptions` dictionary.

Some example reference graphs:

ParseException

![ParseException_reference_graph](https://user-images.githubusercontent.com/5595494/54885125-4e9ff700-4e4f-11e9-8996-59683ce457a8.png)

Dicts

![dict_reference_graph](https://user-images.githubusercontent.com/5595494/54885126-5364ab00-4e4f-11e9-83bd-97130535c7df.png)

Tuple

![tuple_reference_graph](https://user-images.githubusercontent.com/5595494/54885127-552e6e80-4e4f-11e9-8e01-0edc7360754f.png)

I then started reading the source code surrounding `all_exceptions`. I have never used PyParsing directly so I might be wrong, but it looks as though the logic for `DEBUG` is inverted. As far as I can tell, if `DEBUG=False` (hardcoded default) then the `record_exception` function _will_ trigger. If `DEBUG=True` then `record_exception` will not be used.

https://github.com/mozilla/moz-sql-parser/blob/dev/moz_sql_parser/sql_parser.py#L39-L42

```python
DEBUG = False

# ...

if DEBUG:
    debug = (None, None, None)
else:
    debug = (nothing, nothing, record_exception)
```

Shouldn't this be inverted? I only expect to record exceptions if debug is true. Maybe I am missing some details of pyparsing. This logic is the same for `master` and `dev`.

**Comment by erp12** [comment]
I may have open this issue prematurely... I have been doing some experimentation no a local clone and I have a better understanding of what is going on.

I tried enabling `DEBUG` on a local clone and noticed that pyparsing will print all exceptions to stdout. I assume that is why the the pyparsing default is overwritten when `DEBUG` is false.

I tried the following with `DEBUG = False`:

```python
if DEBUG:
    debug = (None, None, None)
else:
    debug = (nothing, nothing, nothing)
```

I got a test failure, and I know that it would break [this](https://github.com/mozilla/moz-sql-parser/blob/dev/moz_sql_parser/__init__.py#L39) however it fixed my memory issue:

![no_record_exception](https://user-images.githubusercontent.com/5595494/54885659-1d2a2a00-4e55-11e9-84c0-70797a039c65.png)

It's possible that my use case is unique. I am working on a programming-by-example AI project that synthesizes sql queries. It requires parsing thousands (maybe millions) of generated sql queries, and thus the cost of accumulating the ParseExceptions is high.

I can understand if this is not something you want to officially address, however it would be nice to somehow make `record_exception` optional.
